
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <title>Tools for multiple sequence alignment &#8212; jMSA 1.0-SNAPSHOT documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="How to score a MSA" href="how_to_score.html" />
    <link rel="prev" title="Installation" href="installation.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="tools-for-multiple-sequence-alignment">
<h1>Tools for multiple sequence alignment<a class="headerlink" href="#tools-for-multiple-sequence-alignment" title="Permalink to this headline">¶</a></h1>
<p>One of the main characteristics that sequence alignment tools present is the type of heuristic algorithm that they implement to achieve their objective. During this document the explanation and differences of each of these techniques will be shown along with certain examples of tools that make use of them.</p>
<p>For all of examples we will use as input the file <em>BB11001.tfa</em> whose content is shown below:</p>
<blockquote>
<div><div class="figure align-default">
<img alt="_images/SecuenciaEjemplo.png" src="_images/SecuenciaEjemplo.png" />
</div>
</div></blockquote>
<div class="section" id="progressive-alignment-construction">
<h2>1. Progressive alignment construction<a class="headerlink" href="#progressive-alignment-construction" title="Permalink to this headline">¶</a></h2>
<p>This is the most widely used approach for multiple sequence alignments. It uses a heuristic search known as a progressive technique, also called a hierarchical or tree method.</p>
<p>Progressive alignment builds up a final MSA by combining pairwise alignments beginning with the most similar pair and progressing to the most distantly related. All progressive alignment methods require two stages: a first stage in which the relationships between the sequences are represented as a tree, called a guide tree, and a second step in which the MSA is built by adding the sequences sequentially to the growing MSA according to the guide tree.</p>
<p>The main problem is that when mistakes are made at any stage of the growth of the MSA, these mistakes propagate to the end result. However, they have the advantage that they are efficient enough to implement them on a large scale for many sequences (from 100 to 1000).</p>
<p>The best-known tool examples for this system are the Clustal family, T-Coffee and PSAlign (semi-progressive):</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>Clustal</strong>: This is the most popular progressive alignment method currently. It makes use of seeded guide trees and HMM profile-profile techniques to generate alignments between three or more sequences. Produces biologically significant multiple sequence alignments of divergent sequences. Evolutionary relationships can be seen through Cladograms or Philograms.</p></li>
</ul>
<div class="figure align-default">
<img alt="_images/ClustalW.png" src="_images/ClustalW.png" />
</div>
<ul class="simple">
<li><p><strong>T-Coffee</strong>: It is slower than Clustal and its derivatives but generally produces more accurate alignments for distantly related sequence sets. T-Coffee calculates pairwise alignments by combining the direct alignment of the pair with indirect alignments that aligns each sequence of the pair to a third sequence. It uses the output from Clustal as well as another local alignment program LALIGN, which finds multiple regions of local alignment between two sequences. The resulting alignment and phylogenetic tree are used as a guide to produce new and more accurate weighting factors.</p></li>
</ul>
<div class="figure align-default">
<img alt="_images/T-Coffee.png" src="_images/T-Coffee.png" />
</div>
<ul class="simple">
<li><p><strong>PSAlign</strong>: Tool that implements a semi-progressive method that improves the quality of the alignment and does not use a lossy heuristic while running in polynomial time.</p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="iterative-methods">
<h2>2. Iterative methods<a class="headerlink" href="#iterative-methods" title="Permalink to this headline">¶</a></h2>
<p>These work similarly to progressive methods but repeatedly realign the initial sequences and add new sequences to the growing MSA. Their main feature is that they can revert to previously calculated pairwise alignments or sub-MSAs that incorporate subsets of the query sequence as a means of optimizing an overall objective function such as finding a high-quality alignment score.</p>
<p>The best-known tools that make use of this approach are PRRN / PRRP, CHAOS / DIALIGN and MUSCLE.</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>PRRN / PRRP</strong>: Uses a scaling algorithm to optimize your MSA alignment score and iteratively corrects both alignment weights and locally diverging or “hollow” regions of the growing MSA. PRRP works best when refining a previously built alignment by a faster method.</p></li>
</ul>
<div class="figure align-default">
<img alt="_images/PRRN.png" src="_images/PRRN.png" />
</div>
<ul class="simple">
<li><p><strong>CHAOS / DIALOG</strong>: Take an unusual approach of focusing closely on local alignments between subsegments or sequence motifs without introducing a gap penalty. Alignment of the individual motifs is accomplished with a matrix representation similar to a dot matrix graph in a pairwise alignment. An alternative method is implemented that uses fast local alignments as anchor points or “seeds” for a slower global alignment procedure.</p></li>
</ul>
<div class="figure align-default">
<img alt="_images/CHAOS+DIALIGN.png" src="_images/CHAOS+DIALIGN.png" />
</div>
<ul class="simple">
<li><p><strong>MUSCLE</strong>: This is a log-expectation multiple sequence alignment that improves progressive methods with a more accurate distance measurement to evaluate the relationship of two sequences. The distance measurement is updated between the iteration stages.</p></li>
</ul>
<div class="figure align-default">
<img alt="_images/MUSCLE.png" src="_images/MUSCLE.png" />
</div>
</div></blockquote>
</div>
<div class="section" id="consensus-methods">
<h2>3. Consensus methods<a class="headerlink" href="#consensus-methods" title="Permalink to this headline">¶</a></h2>
<p>Consensus methods attempt to find the optimal multiple sequence alignment given multiple different alignments of the same set of sequences. There are two commonly used consensus methods:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>M-COFFEE</strong>: It is part of the T-Coffee distribution. The specificity of M-Coffee is that rather than computing a multiple sequence alignment on its own, it uses other packages to compute the alignments. It then uses T-Coffee to combine all these alignments into one unique final alignment. In practice, this means that if you are using several packages to generate your alignments, you can combine these alignments rather than choosing one of them. In practice we have shown that the combined alignments are on average better than the initial alignments. Furthermore, the regions where they agree tend to be correctly aligned.</p></li>
</ul>
<div class="figure align-default">
<img alt="_images/M-Coffee.png" src="_images/M-Coffee.png" />
</div>
<ul class="simple">
<li><p><strong>MergeAlign</strong>: is a program that constructs a consensus multiple sequence alignment from multiple independent alignments. Using dynamic programming it efficiently combines individual multiple sequence alignments to generate a consensus that is maximally representative of all constituent alignments. Using Merge Align to combine multiple sequence alignments generated using different matrices of amino acid substitution produces multiple sequence alignments that are more robust and more accurate than alignments generated using only a single matrix of amino acid substitution.</p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="hidden-markov-models">
<h2>4. Hidden Markov models<a class="headerlink" href="#hidden-markov-models" title="Permalink to this headline">¶</a></h2>
<p>Hidden Markov models are probabilistic models that can assign likelihoods to all possible combinations of gaps, matches, and mismatches to determine the most likely MSA or set of possible MSAs. HMMs can produce both global and local alignments. Although HMM-based methods have been developed relatively recently, they offer significant improvements in computational speed, especially for sequences that contain overlapping regions.</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>POA (Partial Order Alignment)</strong>: it is a fast program for multiple sequence alignment in bioinformatics. Its advantages are speed, scalability, sensitivity, and the superior ability to handle branching / indels in the alignment.</p></li>
<li><p><strong>SAM</strong>: is a collection of flexible software tools for creating, refining, and using linear hidden Markov models for biological sequence analysis. The model states can be viewed as representing the sequence of columns in a multiple sequence alignment, with provisions for arbitrary position-dependent insertions and deletions in each sequence. The models are trained on a family of protein or nucleic acid sequences using an expectation-maximization algorithm and a variety of algorithmic heuristics.</p></li>
<li><p><strong>HMMER</strong>:  Its general usage is to identify homologous protein or nucleotide sequences, and to perform sequence alignments. It detects homology by comparing a profile-HMM to either a single sequence or a database of sequences. Sequences that score significantly better to the profile-HMM compared to a null model are considered to be homologous to the sequences that were used to construct the profile-HMM</p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="phylogeny-aware-methods">
<h2>5. Phylogeny-aware methods<a class="headerlink" href="#phylogeny-aware-methods" title="Permalink to this headline">¶</a></h2>
<p>Most multiple sequence alignment methods try to minimize the number of insertions/deletions (gaps) and, as a consequence, produce compact alignments. This causes several problems if the sequences to be aligned contain non-homologous regions, if gaps are informative in a phylogeny analysis. These problems are common in newly produced sequences that are poorly annotated and may contain frame-shifts, wrong domains or non-homologous spliced exons.
Some tools for this system are the following:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>PRANK</strong>: It’s based on an algorithm that treats insertions correctly and avoids over-estimation of the number of deletion events. In addition, PRANK borrows ideas from maximum likelihood methods used in phylogenetics and correctly takes into account the evolutionary distances between sequences. Lastly, PRANK allows for defining a potential structure for sequences to be aligned and then, simultaneously with the alignment, predicts the locations of structural units in the sequences. Let see an example of this tool:</p></li>
</ul>
<div class="figure align-default">
<img alt="_images/WEBPRANK1.png" src="_images/WEBPRANK1.png" />
</div>
<div class="figure align-default">
<img alt="_images/webprank2.png" src="_images/webprank2.png" />
</div>
<ul class="simple">
<li><p><strong>PAGAN</strong>: It’s based on the phylogeny-aware progressive alignment algorithm and uses graphs to describe the uncertainty in the presence of characters at certain sequence positions. However, graphs also allow describing the uncertainty in input sequences and modelling e.g. homopolymer errors in Roche 454 reads, or representing inferred ancestral sequences against which other sequences can then be aligned.</p></li>
<li><p><strong>ProGraphMSA</strong>: It’s a state-of-the-art multiple sequence alignment tool which produces phylogenetically sensible gap patterns while maintaining robustness by allowing alternative splicings and errors in the branching pattern of the guide tree. This is achieved by incorporating a graph-based sequence representation and combines it with the advantages of the phylogeny-aware algorithm in Prank. Further, we account for variations in the substitution pattern by using estimated amino acid frequencies and by implementing context-specific profiles.</p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="motif-finding">
<h2>6. Motif finding<a class="headerlink" href="#motif-finding" title="Permalink to this headline">¶</a></h2>
<p>A motif is a sequence pattern that occurs repeatedly in a group of related protein or DNA sequences and is often associated with some biological function. Motif finding, also known as profile analysis, is a method of locating sequence motifs in global MSAs that is both a means of producing a better MSA and a means of producing a scoring matrix for use in searching other sequences for similar motifs. A variety of methods for isolating the motifs have been developed, but all are based on identifying short highly conserved patterns within the larger alignment and constructing a matrix similar to a substitution matrix that reflects the amino acid or nucleotide composition of each position in the putative motif. The alignment can then be refined using these matrices. In standard profile analysis, the matrix includes entries for each possible character as well as entries for gaps. Alternatively, statistical pattern-finding algorithms can identify motifs as a precursor to an MSA rather than as a derivation. In many cases when the query set contains only a small number of sequences or contains only highly related sequences, pseudocounts are added to normalize the distribution reflected in the scoring matrix.</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>MEME(Multiple EM for Motif Elicitation)</strong>: is a tool for discovering motifs in a group of related DNA or protein sequences. It’s represents motifs as position-dependent letter-probability matrices which describe the probability of each possible letter at each position in the pattern. Individual MEME motifs do not contain gaps. Patterns with variable-length gaps are split by MEME into two or more separate motifs. This tool takes as input a group of DNA or protein sequences (the training set) and outputs as many motifs as requested. It uses statistical modeling techniques to automatically choose the best width, number of occurrences, and description for each motif.</p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="non-coding-multiple-sequence-alignment">
<h2>7. Non-coding multiple sequence alignment<a class="headerlink" href="#non-coding-multiple-sequence-alignment" title="Permalink to this headline">¶</a></h2>
<p>Non-coding DNA regions, especially TFBSs (Transcription factor binding site), are rather more conserved and not necessarily evolutionarily related, and may have converged from non-common ancestors. Thus, the assumptions used to align protein sequences and DNA coding regions are inherently different from those that hold for TFBS sequences. Although it is meaningful to align DNA coNon-coding DNA regions, especially TFBSs, are rather more conserved and not necessarily evolutionarily related, and may have converged from non-common ancestors. Thus, the assumptions used to align protein sequences and DNA coding regions are inherently different from those that hold for TFBS sequences. Although it is meaningful to align DNA coding regions for homologous sequences using mutation operators, alignment of binding site sequences for the same transcription factor cannot rely on evolutionary related mutation operations. Similarly, the evolutionary operator of point mutations can be used to define an edit distance for coding sequences, but this has little meaning for TFBS sequences because any sequence variation has to maintain a certain level of specificity for the binding site to function. This becomes specifically important when trying to align known TFBS sequences to build supervised models to predict unknown locations of the same TFBS. Hence, Multiple Sequence Alignment methods need to adjust the underlying evolutionary hypothesis and the operators used as in the work published incorporating neighbouring base thermodynamic information regions for homologous sequences using mutation operators, alignment of binding site sequences for the same transcription factor cannot rely on evolutionary related mutation operations.</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>EDNA (Energy Based Multiple Sequence Alignment)</strong>: It is a multi-sequence alignment (MSA) program to align sequences of transcription factor binding sites (TFBS). The novelty of this software is scoring using a thermodynamically generated null hypothesis. The method is well suited for aligning often unrelated sequences.</p></li>
</ul>
</div></blockquote>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">jMSA</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Index:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="problem.html">Multiple sequence alignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tools for multiple sequence alignment</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#progressive-alignment-construction">1. Progressive alignment construction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#iterative-methods">2. Iterative methods</a></li>
<li class="toctree-l2"><a class="reference internal" href="#consensus-methods">3. Consensus methods</a></li>
<li class="toctree-l2"><a class="reference internal" href="#hidden-markov-models">4. Hidden Markov models</a></li>
<li class="toctree-l2"><a class="reference internal" href="#phylogeny-aware-methods">5. Phylogeny-aware methods</a></li>
<li class="toctree-l2"><a class="reference internal" href="#motif-finding">6. Motif finding</a></li>
<li class="toctree-l2"><a class="reference internal" href="#non-coding-multiple-sequence-alignment">7. Non-coding multiple sequence alignment</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="how_to_score.html">How to score a MSA</a></li>
<li class="toctree-l1"><a class="reference internal" href="scores.html">Scores</a></li>
<li class="toctree-l1"><a class="reference internal" href="substitutionMatrix.html">Subtitution Matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="utils.html">Utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="installation.html" title="previous chapter">Installation</a></li>
      <li>Next: <a href="how_to_score.html" title="next chapter">How to score a MSA</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, PAB2020.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.0.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/alignment_tools.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>