
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <title>Scores &#8212; jMSA 1.0-SNAPSHOT documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Subtitution Matrix" href="substitutionMatrix.html" />
    <link rel="prev" title="How to score a MSA" href="how_to_score.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="scores">
<h1>Scores<a class="headerlink" href="#scores" title="Permalink to this headline">¶</a></h1>
<div class="section" id="star">
<h2>Star<a class="headerlink" href="#star" title="Permalink to this headline">¶</a></h2>
<p>The Star method is one of the algorithms used to score multi-sequence alignment.
It was proposed by Gusfield in 1997 and consists of finding the best aligned sequence
with the others (center) by scoring in pairs. This method makes use of a substitution matrix.</p>
<p>The following process is carried out:</p>
<ul class="simple">
<li><p>Search for the most recurring symbol (nucleotide or aminoacid) in each column of the sequence set.</p></li>
<li><p>Calculate the distances from the remaining symbols to the most repeated from a substitution matrix.</p></li>
</ul>
<div class="figure align-default">
<img alt="_images/Captura2.PNG" src="_images/Captura2.PNG" />
</div>
<p>That way, the form of a Star emerges:</p>
<div class="figure align-default">
<img alt="_images/Captura4.PNG" src="_images/Captura4.PNG" />
</div>
<p>IMPLEMENTATION</p>
<p>We have created a function called <cite>aaMoreRepeated</cite>, in which we select the most repeated symbol
of all the sequences in each column, so we’re going build a map in which we will store the symbol
along with its frequency of occurrence.</p>
<p>For every character in the column we’re working with, we check if it’s
included on the map. If it is not, we include it as a key and put its value to 1.
If it is included, we add one to the frequency, delete the previous relationship
and introduce the amino acid with its new frequency.</p>
<div class="highlight-java notranslate"><div class="highlight"><pre><span></span><span class="kd">public</span> <span class="n">Character</span> <span class="nf">aaMoreRepeated</span><span class="p">(</span><span class="kt">char</span><span class="o">[][]</span> <span class="n">parameterSequence</span><span class="p">,</span> <span class="kt">int</span> <span class="n">parameterPosAAinSequence</span><span class="p">){</span>

    <span class="n">HashMap</span><span class="o">&lt;</span><span class="n">Character</span><span class="p">,</span> <span class="n">Integer</span><span class="o">&gt;</span> <span class="n">charFreq</span> <span class="o">=</span> <span class="k">new</span> <span class="n">HashMap</span><span class="o">&lt;</span><span class="n">Character</span><span class="p">,</span> <span class="n">Integer</span><span class="o">&gt;</span><span class="p">();</span>

    <span class="k">for</span> <span class="p">(</span><span class="kt">int</span> <span class="n">numberOfSequence</span> <span class="o">=</span> <span class="mi">0</span><span class="p">;</span> <span class="n">numberOfSequence</span> <span class="o">&lt;</span> <span class="n">parameterSequence</span><span class="p">.</span><span class="na">length</span><span class="p">;</span> <span class="n">numberOfSequence</span><span class="o">++</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">Character</span> <span class="n">aa</span> <span class="o">=</span> <span class="n">parameterSequence</span><span class="o">[</span><span class="n">numberOfSequence</span><span class="o">][</span><span class="n">parameterPosAAinSequence</span><span class="o">]</span><span class="p">;</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">charFreq</span><span class="p">.</span><span class="na">containsKey</span><span class="p">(</span><span class="n">parameterSequence</span><span class="o">[</span><span class="n">numberOfSequence</span><span class="o">][</span><span class="n">parameterPosAAinSequence</span><span class="o">]</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">Integer</span> <span class="n">newFreq</span> <span class="o">=</span> <span class="p">(</span><span class="n">Integer</span><span class="p">)</span> <span class="n">charFreq</span><span class="p">.</span><span class="na">get</span><span class="p">(</span><span class="n">aa</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">;</span>
        <span class="n">charFreq</span><span class="p">.</span><span class="na">remove</span><span class="p">(</span><span class="n">aa</span><span class="p">);</span>
        <span class="n">charFreq</span><span class="p">.</span><span class="na">put</span><span class="p">(</span><span class="n">aa</span><span class="p">,</span> <span class="n">newFreq</span><span class="p">);</span>
      <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">charFreq</span><span class="p">.</span><span class="na">put</span><span class="p">(</span><span class="n">aa</span><span class="p">,</span> <span class="mi">1</span><span class="p">);</span>
      <span class="p">}</span>
    <span class="p">}</span>
</pre></div>
</div>
<p>Once all the frequencies have been calculated, we must select the highest one.
We use a for loop in which we compare the frequency saved for each character
with a counter that we’ve initialized to 0. Every time the frequency is higher
than the one stored in the counter, we update the counter with that new frequency
and we store the amino acid that has that frequency.</p>
<div class="highlight-java notranslate"><div class="highlight"><pre><span></span><span class="kt">int</span> <span class="n">moreRepeatedFreq</span> <span class="o">=</span> <span class="mi">0</span><span class="p">;</span>
<span class="n">Character</span> <span class="n">moreRepeatedAA</span> <span class="o">=</span> <span class="n">parameterSequence</span><span class="o">[</span><span class="mi">0</span><span class="o">][</span><span class="n">parameterPosAAinSequence</span><span class="o">]</span><span class="p">;</span>

<span class="k">for</span> <span class="p">(</span><span class="n">Map</span><span class="p">.</span><span class="na">Entry</span><span class="o">&lt;</span><span class="n">Character</span><span class="p">,</span> <span class="n">Integer</span><span class="o">&gt;</span> <span class="n">entry</span> <span class="p">:</span> <span class="n">charFreq</span><span class="p">.</span><span class="na">entrySet</span><span class="p">())</span> <span class="p">{</span>
  <span class="k">if</span> <span class="p">(</span><span class="n">entry</span><span class="p">.</span><span class="na">getValue</span><span class="p">()</span> <span class="o">&gt;</span> <span class="n">moreRepeatedFreq</span><span class="p">)</span> <span class="p">{</span>
    <span class="n">moreRepeatedAA</span> <span class="o">=</span> <span class="n">entry</span><span class="p">.</span><span class="na">getKey</span><span class="p">();</span>
    <span class="n">moreRepeatedFreq</span> <span class="o">=</span> <span class="n">entry</span><span class="p">.</span><span class="na">getValue</span><span class="p">();</span>
  <span class="p">}</span>
<span class="p">}</span>
<span class="k">return</span> <span class="n">moreRepeatedAA</span><span class="p">;</span>
</pre></div>
</div>
<p>Finally, we calculate the distances of each symbol in the column from the
symbol with the highest value in that column. We do this within the function
called <cite>compute</cite>. This function is introduced by the set of sequences.
We go through each of the columns, and for each one we call the function ‘moreRepeatedAA’,
which will give us the most repeated character.</p>
<p>Once we have it, we calculate the distance of that symbol with the rest
of the symbols of the column.</p>
<div class="highlight-java notranslate"><div class="highlight"><pre><span></span><span class="kd">public</span> <span class="kt">double</span> <span class="nf">compute</span><span class="p">(</span><span class="kt">char</span><span class="o">[][]</span> <span class="n">sequence</span><span class="p">)</span> <span class="p">{</span>
    <span class="kt">double</span> <span class="n">result</span> <span class="o">=</span> <span class="mi">0</span><span class="p">;</span>

    <span class="k">for</span> <span class="p">(</span><span class="kt">int</span> <span class="n">posAAinSequence</span> <span class="o">=</span> <span class="mi">0</span><span class="p">;</span> <span class="n">posAAinSequence</span> <span class="o">&lt;</span> <span class="n">sequence</span><span class="o">[</span><span class="mi">0</span><span class="o">]</span><span class="p">.</span><span class="na">length</span><span class="p">;</span> <span class="n">posAAinSequence</span><span class="o">++</span><span class="p">)</span> <span class="p">{</span>

      <span class="n">Character</span> <span class="n">moreRepeatedAA</span> <span class="o">=</span> <span class="n">aaMoreRepeated</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">posAAinSequence</span><span class="p">);</span>

      <span class="c1">// 2. We compute the punctuation for this column</span>
      <span class="k">for</span> <span class="p">(</span><span class="kt">char</span><span class="o">[]</span> <span class="n">chars</span> <span class="p">:</span> <span class="n">sequence</span><span class="p">)</span> <span class="p">{</span>
        <span class="kt">double</span> <span class="n">distance</span> <span class="o">=</span> <span class="n">scoreMatrix</span><span class="p">.</span><span class="na">getDistance</span><span class="p">(</span><span class="n">moreRepeatedAA</span><span class="p">,</span> <span class="n">chars</span><span class="o">[</span><span class="n">posAAinSequence</span><span class="o">]</span><span class="p">);</span>

        <span class="c1">// 3. We add the distance to the total result we have to return</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">result</span> <span class="o">+</span> <span class="n">distance</span><span class="p">;</span>
      <span class="p">}</span>
    <span class="p">}</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">result</span><span class="p">);</span>
  <span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="sum-of-pairs">
<h2>Sum of pairs<a class="headerlink" href="#sum-of-pairs" title="Permalink to this headline">¶</a></h2>
<p>Sum of pairs is a method of evaluating the quality of multiple sequence alignment.
This method assigns a specific score to the set of possible pair combinations of elements using a substitution matrix. Its operation consists of, given a sequence alignment, selecting by column the possible pair combinations and identifying the sum of the values provided by the substitution matrix, which gives us a final sum made up of the sums of each column.</p>
<dl>
<dt>That is:</dt><dd><ul class="simple">
<li><p>Given an alignment A composed of 4 sequences such that:</p></li>
</ul>
<div class="figure align-default">
<img alt="_images/SecuenciaDocumentacion.PNG" src="_images/SecuenciaDocumentacion.PNG" />
</div>
<ul class="simple">
<li><p>And a substitution matrix such that:</p></li>
</ul>
<div class="figure align-default">
<img alt="_images/MatrizDocumentacion.PNG" src="_images/MatrizDocumentacion.PNG" />
</div>
</dd>
</dl>
<p>Our method sequentially selects each column and for each column it will select all possible nucleotide pair combinations for that column. Each possible pair will be translated into a numerical value given by the substitution matrix, these values will be added together to finally obtain a final value.
The higher the final value, the better the alignment.</p>
<p>In the example of our images, for the column marked in red, the sum of pairs would give us the following result.</p>
<div class="figure align-default">
<img alt="_images/ResultadoDocumentacion.PNG" src="_images/ResultadoDocumentacion.PNG" />
</div>
<p>To carry out the sum of pairs in java we have created two methods: compute and sumCol.</p>
<p>The sumCol method will receive as parameters the sequence and the column in question and will make the sum of pairs of that column.
For this, it consists of two nested for loops: the first will go through the column from the first element to the last (these elements will correspond to the first element of the pair); and the second for that will go through the elements of the column starting from the element already selected in the previous for and until the last one (these elements will correspond to the second element of the pair).
In this way all possible pairs are selected.
Once the pair is selected, the sumCol method calls the getDistance function of the Substitution matrix class, which is responsible for finding the value corresponding to that pair.</p>
<div class="figure align-default">
<img alt="_images/sumColDocumentacion.PNG" src="_images/sumColDocumentacion.PNG" />
</div>
<p>On the other hand, the compute method receives the sequence as a parameter and is in charge of making the sum of all the columns.
To do this, it goes through the sequence through columns using a for and calls the sumCol method, passing the column in question as a parameter, which will return the sum of that column.</p>
<div class="figure align-default">
<img alt="_images/computeDocumentacion.PNG" src="_images/computeDocumentacion.PNG" />
</div>
</div>
<div class="section" id="percentage-of-non-gaps">
<h2>Percentage of non-gaps<a class="headerlink" href="#percentage-of-non-gaps" title="Permalink to this headline">¶</a></h2>
<p>Visual depictions of the alignment illustrate mutation events such as point mutations (single amino acid or nucleotide changes) that appear as differing characters in a single alignment column, and insertion or deletion mutations (indels or gaps) that appear as hyphens.</p>
<p>This part of the programme is going to calculate the number of non-gaps in an alignment sequence. This is super important since the objective of the MSA, among others,  is to optimize to the maximum the value of the non-gaps-percentage.</p>
<p>That is the reason why we are going to create the class <em>PercentageOfNonGaps</em></p>
<p>This class implements Score class. It has a method called <em>compute()</em> which we pass a matrix as a parameter. A counter for non-gaps is initialized to zero This method has nested loop to loop through it.  If the element we read is different from a hyphens, we increase the counter.
Finally we divide this counter between the total length of the sequence. To obtain a percentage over 100, we multiply the result by 100.</p>
</div>
<div class="section" id="percentage-of-totally-conserved-columns">
<h2>Percentage of totally conserved columns<a class="headerlink" href="#percentage-of-totally-conserved-columns" title="Permalink to this headline">¶</a></h2>
<dl>
<dt>The percentage of conserved columns is very important because it has many aplications in studies of phylogenetics and taxonomy, in medical researches and in functional annotation.</dt><dd><blockquote>
<div><ul class="simple">
<li><p>Phylogenetics and taxonomy: the sets of conserved columns can be used to create an evolution tree to study the species and their ancestors.</p></li>
<li><p>In medical researches the applications of the MSA and specifically the amount of conserved columns it’s used to identifi genetics diseases, in the compatibility between species so we can study the pathology on lab organisms instead of humans…</p></li>
<li><p>Identifying the conserved columns can be very helpful to predict functional sequences such as genes, besides, if the secuences have a known function, they can also be used to predict the function of those proteins; in order to create databases.</p></li>
</ul>
</div></blockquote>
<p>Our class measures the amount of conserved columns by implementing the class score, and finding the sequences who are conserved and then returning the percentege ot total conserved columns of the array.</p>
</dd>
</dl>
</div>
<div class="section" id="entropy">
<h2>Entropy<a class="headerlink" href="#entropy" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Definition:</p></li>
</ul>
<p>Entropy of Shanon can be defined as grade of disorder that a random variable can generate. It is related to the uncertainty that exists in
an experiment, as the amount of “noise” that contains the system.</p>
<p>However, it is also related to  the physics’ magnitude that allows us to measure the non-usable part of the energy
contained in a system of particles or data. That means that that part of the total can not be used to produce
physical work.</p>
<p>In our case, it is useful in order to measure how diverse are the residues in a column. This is vey important when scoring MSA because it
represents how much distance exist between the sequences. Entropy is important because it tells us that systems can’t go back in their processes and how is the energy wasted in a closed environment.
Mathematically it’s represented by the summation of residue’s frequency times the logarithm of different residue’s number, as it can be observed in
the following picture:</p>
<img alt="_images/entropy.jpg" src="_images/entropy.jpg" />
<ul class="simple">
<li><p>Implementation:</p></li>
</ul>
<p>We have created a function called compute, which receives a sequence and creates a HashMap &lt;String, Integer&gt;
with the number of times a nucleotide appears. We have supposed that the sequence doesn’t contains incorrect characters.
Once we have the frequencies of nucleotide, we can apply the formula of Shanon Entropy.
In addition, we have implemented a range of JUnit Tests in order to be secure that the code works correctly.</p>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">jMSA</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Index:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="problem.html">Multiple sequence alignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="alignment_tools.html">Tools for multiple sequence alignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="how_to_score.html">How to score a MSA</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Scores</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#star">Star</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sum-of-pairs">Sum of pairs</a></li>
<li class="toctree-l2"><a class="reference internal" href="#percentage-of-non-gaps">Percentage of non-gaps</a></li>
<li class="toctree-l2"><a class="reference internal" href="#percentage-of-totally-conserved-columns">Percentage of totally conserved columns</a></li>
<li class="toctree-l2"><a class="reference internal" href="#entropy">Entropy</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="substitutionMatrix.html">Subtitution Matrix</a></li>
<li class="toctree-l1"><a class="reference internal" href="utils.html">Utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="how_to_score.html" title="previous chapter">How to score a MSA</a></li>
      <li>Next: <a href="substitutionMatrix.html" title="next chapter">Subtitution Matrix</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, PAB2020.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.0.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/scores.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>